packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx

```diff
@@ -0,0 +1,192 @@
+import MaskbookPluginWrapper from '../../MaskbookPluginWrapper'
+import { usePostInfoDetails } from '../../../components/DataSource/usePostInfo'
+import { Button, Box, Typography, Select, MenuItem, FormControl, CircularProgress } from '@material-ui/core'
+import { useI18N } from '../../../utils/i18n-next-ui'
+import { init, changeMediaDevice, disconnectVoice, setMuted, registerSpeakingUsernames } from '../utils/voicechat'
+import type { PeerMediaElement, VoiceChatMetadata } from '../types'
+import { VolumeUpOutlined, VolumeMuteOutlined, VolumeOffOutlined, CloseOutlined } from '@material-ui/icons'
+import { useState } from 'react'
+import { PortalShadowRoot } from '../../../utils/shadow-root/ShadowRootPortal'
+
+interface VoicechatInlayProps {
+    metadata: VoiceChatMetadata
+}
+
+export function VoicechatInlay(props: VoicechatInlayProps) {
+    const postIdentifier = usePostInfoDetails('postIdentifier')
+    const postBy = usePostInfoDetails('postBy')
+
+    const [peerMediaElements, setPeerMediaElements] = useState<Record<string, PeerMediaElement>>()
+    const [localPeerMediaElement, setLocalPeerMediaElement] = useState<PeerMediaElement | null>()
+    const [mediaDevices, setMediaDevices] = useState<MediaDeviceInfo[]>([])
+    const [activeMediaDevice, setActiveMediaDevice] = useState<string>('')
+    const [isMuted, setIsMuted] = useState<boolean>(false)
+    const [isJoining, setIsJoining] = useState<boolean>(false)
+
+    const { t } = useI18N()
+
+    if (!postIdentifier) {
+        return null
+    }
+
+    const joinChat = () => {
+        setIsJoining(true)
+
+        navigator.mediaDevices.enumerateDevices().then((devices) => {
+            setMediaDevices([...devices.filter((d) => d.kind === 'audioinput')])
+        })
+
+        const customServer = props.metadata.customServer ? props.metadata.customServer : ''
+
+        init(
+            customServer,
+            postIdentifier.postId,
+            postBy.userId,
+            (peerMedia, localPeer) => {
+                setPeerMediaElements({ ...peerMedia })
+                setLocalPeerMediaElement(localPeer)
+                setIsJoining(false)
+            },
+            (activeMediaDevice: string) => {
+                setActiveMediaDevice(activeMediaDevice)
+            },
+        )
+    }
+
+    const JoinButton = () => {
+        return localPeerMediaElement ? null : (
+            <Button
+                color="primary"
+                variant="contained"
+                style={{ color: '#fff' }}
+                onClick={() => joinChat()}
+                startIcon={isJoining ? <CircularProgress style={{ color: 'white' }} size={24} /> : null}>
+                {t('plugin_voicechat_join_conversation')}
+            </Button>
+        )
+    }
+
+    const ChannelUsers = () => {
+        let elements: PeerMediaElement[] = []
+
+        const [speakingUsernames, setSpeakingUsernames] = useState<string[]>([])
```

Comment:
Please do not define a new component inside another component. It will cause the component tree to be totally destroyed and the state is lost.